File: src/cs2/aimbot.cpp
#include "cs2/cs2.hpp"
#include "cs2/features.hpp"
#include "math.hpp"
#include "mouse.hpp"

void Aimbot() {
    const std::optional<Player> local_player = Player::LocalPlayer();
    if (!local_player) {
        return;
    }

    const WeaponConfig &aim_config = config.aimbot.CurrentWeaponConfig(local_player->WeaponName());

    if (!config.aimbot.global.enabled || !target.player || !IsButtonPressed(config.aimbot.hotkey)) {
        return;
    }

    if (!target.player->IsValid()) {
        return;
    }

    if (aim_config.flash_check && local_player->IsFlashed()) {
        return;
    }

    if (aim_config.visibility_check) {
        if ((target.player->SpottedMask() & 1 << target.local_pawn_index) == 0) {
            return;
        }
    }

    glm::vec2 target_angle {};
    if (aim_config.multibone) {
        target_angle = target.angle;
    } else {
        target_angle = TargetAngle(
            local_player->EyePosition(), target.player->BonePosition(Bones::Head), aim_punch);
    }

    const glm::vec2 view_angles = local_player->ViewAngles();
    if (AnglesToFov(view_angles, target_angle) > aim_config.fov * DistanceScale(target.distance)) {
        return;
    }

    if (local_player->ShotsFired() < aim_config.start_bullet) {
        return;
    }

    glm::vec2 aim_angles {view_angles - target_angle};
    if (aim_angles.y < -180.0f) {
        aim_angles.y += 360.0f;
    }
    Vec2Clamp(aim_angles);

    const f32 sensitivity = Sensitivity() * local_player->FovMultiplier();

    const glm::vec2 xy {aim_angles.y / sensitivity * 25.0f, -aim_angles.x / sensitivity * 25.0f};
    glm::vec2 smooth_angles;
    if (!aim_config.aim_lock && aim_config.smooth > 0.0f) {
        smooth_angles =
            glm::vec2 {xy.x / (aim_config.smooth + 1.0f), xy.y / (aim_config.smooth + 1.0f)};
    } else {
        smooth_angles = xy;
    }

    const glm::ivec2 smooth_int(
        static_cast<i32>(smooth_angles.x), static_cast<i32>(smooth_angles.y));
    MouseMove(smooth_int);
}

File: src/cs2/aimbot.rs
use glam::Vec2;
use crate::{
    cs2::{CS2, entity::player::Player},
    math::{angles_to_fov, vec2_clamp},
    os::mouse::Mouse,
};

// ---------- helpers ---------------------------------------------------------
fn human_curve(x: f32) -> f32 { x * x * (3.0 - 2.0 * x) }

fn white(seed: u32) -> f32 {
    let s = (seed << 13) ^ seed;
    let s = s.wrapping_mul(s.wrapping_mul(15731).wrapping_add(789221));
    ((s >> 16) & 0x7fff) as f32 / 16384.0 - 1.0
}

// ---------- aimbot ----------------------------------------------------------
impl CS2 {
    pub fn aimbot(&mut self, _config: &crate::config::Config, mouse: &mut Mouse) {
        // pull the same config you already expose
        let cfg = self.aimbot_config(_config);
        if !cfg.enabled || self.target.player.is_none() { return; }
        let target = self.target.player.as_ref().unwrap();

        let local = match Player::local_player(self) { Some(p) => p, None => return };

        if cfg.flash_check   && local.is_flashed(self)   { return; }
        if cfg.visibility_check && !target.visible(self, &local) { return; }
        if !target.is_valid(self) { return; }
        if local.shots_fired(self) < cfg.start_bullet { return; }

        // best bone
        let (target_angle, _) = cfg.bones.iter()
            .map(|b| target.bone_position(self, b.u64()))
            .map(|p| self.angle_to_target(&local, &p, &self.target.previous_aim_punch))
            .map(|a| (a, angles_to_fov(&local.view_angles(self), &a)))
            .fold((Vec2::ZERO, 360.0), |best, cur| if cur.1 < best.1 { cur } else { best });

        let view = local.view_angles(self);
        let max_fov = cfg.fov * if cfg.distance_adjusted_fov {
            self.distance_scale(self.target.distance) } else { 1.0 };
        if angles_to_fov(&view, &target_angle) > max_fov { return; }

        // humanised move
        let mut aim = view - target_angle;
        if aim.y < -180.0 { aim.y += 360.0; }
        vec2_clamp(&mut aim);

        let sens = self.get_sensitivity() * local.fov_multiplier(self);
        let base = Vec2::new(aim.y / sens * 50.0, -aim.x / sens * 50.0);

        let tick = self.global_tick_count() as u32;
        let jitter = Vec2::new(white(tick), white(tick.wrapping_mul(7919))) * 0.35;
        let slip  = Vec2::new(white(tick.wrapping_mul(13001)), white(tick.wrapping_mul(19717))) * 0.20;

        let smooth = (cfg.smooth + 1.0).clamp(1.0, 20.0);
        let move_ = (base + jitter + slip) * human_curve(1.0 / smooth);

        // anti-twitch: skip 2 frames after big snap
        if self.aimbot_cooldown == 0 {
            mouse.move_rel(&move_);
            if move_.length_squared() > 0.02 { self.aimbot_cooldown = 2; }
        } else {
            self.aimbot_cooldown -= 1;
        }
    }
}
